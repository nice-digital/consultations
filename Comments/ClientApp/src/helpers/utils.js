// @flow

import queryString from "query-string";
//import stringifyObject from "stringify-object";

/**
 * Create a query string out of an object
 * @param obj the key-value pairs that you'd like to be turned into a querystring
 * @returns {string}
 */
export function objectToQueryString(obj) {
	let str = [];
	for (let p in obj)
		if (Object.prototype.hasOwnProperty.call(obj, p)) {
			if (Array.isArray(obj[p])) {
				for (let index in obj[p]) {
					str.push(encodeURIComponent(p) + "=" + encodeURIComponent(obj[p][index]));
				}
			} else {
				str.push(encodeURIComponent(p) + "=" + encodeURIComponent(obj[p]));
			}
		}
	if (str.length) return "?" + str.join("&");
	return "";
}

/**
 * Creates an object out of a query string
 * 	@param qryString {string} the querystring you'd like to be turned into an object
 *  @returns obj
 */
export function queryStringToObject(qryString) {
	return queryString.parse(qryString);
}

/**
 * Resolve a promise on the next tick - used in testing
 * @see https://nodejs.org/en/docs/guides/event-loop-timers-and-nexttick/#process-nexttick
 * @returns {Promise<any>}
 */
export const nextTick = async () => {
	return new Promise((resolve) => {
		setTimeout(resolve, 0);
	});
};

/**
 * Format a string by replacing {0} with arguments passed in
 * @see https://stackoverflow.com/a/4673436
 * @returns {String}
 */
export function replaceFormat(stringToReplace, args) {
	if (typeof (stringToReplace) === "undefined" || stringToReplace === null)
		return stringToReplace;

	return stringToReplace.replace(/{(\d+)}/g, function (match, number) {
		return typeof args[number] !== "undefined"
			? args[number]
			: match
		;
	});
}

/**
 * @param link String The string in which you want to check for the presence of "http"
 * Detects link that starts with "http"
 * @returns {Boolean}
 */
export const isHttpLink = link => link.indexOf("http") === 0;

// Remove querystring parameters from a given URL
export const removeQueryParameter =
	(url: string, parameter: string, value: ?string = null): string => {
		const urlParts = url.split("?");

		if (urlParts.length < 2)
			return url;

		let urlParams = urlParts[1].split(/[&;]/g);

		for (let i = urlParams.length; i-- > 0;) {
			const parParts = urlParams[i].split("=");

			if (parParts[0].toLowerCase() === parameter.toLowerCase()
				&& (!value || (value && value.toLowerCase() === (parParts[1] || "").toLowerCase()))) {
				urlParams.splice(i, 1);
			}
		}

		const queryString = (urlParams.length > 0 ? "?" + urlParams.join("&") : "");
		return urlParts[0] + queryString;
	};

export const stripMultipleQueries = (path, queries) => {
	let strippedPath = path;

	queries.forEach((query) => {
		strippedPath = removeQueryParameter(strippedPath, query);
	});

	return strippedPath;
};

export const appendQueryParameter =
	(url: string, parameter: string, value: string): string => `${url}${url.indexOf("?") === -1 ? "?" : "&"}${parameter}=${value}`;

/// Determines if the execution environment is client or server
/// See https://stackoverflow.com/a/32598826/486434
export const canUseDOM = (): boolean => (
	!!(
		typeof window !== "undefined" &&
		window.document &&
		window.document.createElement
	));

const whichChild = (elem) => {
	let i = 0;
	while ((elem == elem.previousSibling) != null) ++i;
	return i;
};

const reverseString = (str) => {
	return (str === "") ? "" : reverseString(str.substr(1)) + str.charAt(0);
};

//this function returns a long dotted decimal representing the position in the document. below the div with id of root.
export const getElementPositionWithinDocument = (elem) => {
	let curindex = "";
	if (elem.parentElement) {
		do {
			curindex += whichChild(elem).toString() + " "; // add elem.tagName to see what the element is.
			if (elem.id === "root") {
				break;
			}
		} while (elem = elem.parentElement); // eslint-disable-line
	}
	return reverseString(curindex.trim().replace(new RegExp(/\s/, "g"), "."));
};

//this gets the title of the nearest section above the passed in element (or the element itself). it can also return the chapter title.
export const getSectionTitle = (elem) => {
	do {
		if (elem.classList.contains("section")) {
			return elem.title;
		}
		if (elem.classList.contains("chapter")) {
			return elem.title;
		}
		if (elem.id === "root") { //it won't look higher than root.
			return null;
		}
	} while (elem = elem.parentNode); // eslint-disable-line
	return null; //shouldn't really ever be called, as root will get hit if there's no matches.
};

export const removeQuerystring = (url) => {
	return url.split("?")[0];
};
