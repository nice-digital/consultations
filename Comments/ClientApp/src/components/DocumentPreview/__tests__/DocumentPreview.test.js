/* global jest */

import React from "react";
import { shallow, mount } from "enzyme";
import axios from "axios";
import MockAdapter from "axios-mock-adapter";
import { MemoryRouter } from "react-router";

import { DocumentPreview } from "../DocumentPreview";
import PreviewChapterData from "./PreviewChapter.json";
import PreviewConsultationData from "./PreviewConsultation.json";
import PreviewDocumentsData from "./PreviewDocuments.json";
import { nextTick } from "../../../helpers/utils";
import toJson from "enzyme-to-json";

jest.mock("../../../context/UserContext", () => {
	return {
		UserContext: {
			Consumer: (props) => {
				return props.children({
					isAuthorised: true,
				});
			},
		},
	};
});

describe("[ClientApp] ", () => {
	describe("DocumentPreview Component", () => {
		const fakeProps = {
			location: {},
			match: {
				params: {
					consultationId: 113,
					documentId: 1,
					chapterSlug: "draftappendicesa-tcleandocx",
					reference: "GID-NG10186",
				},
			},
			staticContext: {
				preload: {
					data:{
						isAuthorised: true,
					},
				},
			},
		};

		it("should render the loading message", () => {
			const wrapper = shallow(<DocumentPreview {...fakeProps} />, {
				disableLifecycleMethods: true,
			});
			expect(wrapper.find("h1").text()).toEqual("Loading...");
		});

		it("should match snapshot with supplied data", () => {
			const mock = new MockAdapter(axios);

			let documentsPromise = new Promise(resolve => {
				mock
					.onGet("consultations/api/PreviewDraftDocuments?consultationId=113&documentId=1&reference=GID-NG10186")
					.reply(() => {
						resolve();
						return [200, PreviewDocumentsData];
					});
			});

			let consulatationPromise = new Promise(resolve => {
				mock
					.onGet("/consultations/api/DraftConsultation?consultationId=113&documentId=1&reference=GID-NG10186")
					.reply(() => {
						resolve();
						return [200, PreviewConsultationData];
					});
			});

			let chapterPromise = new Promise(resolve => {
				mock
					.onGet(
						"/consultations/api/PreviewChapter?consultationId=113&documentId=1&chapterSlug=draftappendicesa-tcleandocx&reference=GID-NG10186",
					)
					.reply(() => {
						resolve();
						return [200, PreviewChapterData];
					});
			});

			const wrapper = mount(
				<MemoryRouter>
					<DocumentPreview {...fakeProps} />
				</MemoryRouter>,
			);

			return Promise.all([
				documentsPromise,
				consulatationPromise,
				chapterPromise,
			]).then(async () => {
				await nextTick();
				wrapper.update();
				expect(
					toJson(wrapper, {
						noKey: true,
						mode: "deep",
					}),
				).toMatchSnapshot();
			});
		});

		it("getPreviewDocumentChapterLinks method", () => {
			const documentPreview = new DocumentPreview();
			expect(
				documentPreview.getPreviewDocumentChapterLinks(1, "draftappendicesa-tcleandocx", PreviewDocumentsData, "this is my title", "GID-1234", 1)).toHaveProperty("links", [
				{
					"current": true,
					"isReactRoute": true,
					"label": "DRAFT_Appendices_A-T_CLEAN.docx",
					"url": "/preview/GID-1234/consultation/1/document/1/chapter/draftappendicesa-tcleandocx",
				},
				{
					"current": false,
					"isReactRoute": true,
					"label": "Appendices",
					"url": "/preview/GID-1234/consultation/1/document/1/chapter/appendices",
				},
				{
					"current": false,
					"isReactRoute": true,
					"label": "1 Scope",
					"url": "/preview/GID-1234/consultation/1/document/1/chapter/scope",
				}]);
		});

	});
});
