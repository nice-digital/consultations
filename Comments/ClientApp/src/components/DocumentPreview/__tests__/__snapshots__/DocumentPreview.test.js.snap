// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[ClientApp]  DocumentPreview Component should match snapshot with supplied data 1`] = `
Array [
  "",
  <div
    className="container"
  >
    <div
      className="grid"
    >
      <div
        data-g="12"
      >
        <nav
          aria-label="Breadcrumbs"
        >
          <p
            className="visually-hidden"
          >
            You are here:
          </p>
          <ol
            className="breadcrumbs"
            itemScope={true}
            itemType="http://schema.org/BreadcrumbList"
          >
            <li
              className="breadcrumbs__crumb"
              itemProp="itemListElement"
              itemScope={true}
              itemType="http://schema.org/ListItem"
              onClick={[Function]}
              role="presentation"
            >
              <a
                href="/guidance/inconsultation"
                itemProp="item"
              >
                <span
                  itemProp="name"
                >
                  All consultations
                </span>
              </a>
              <meta
                content={0}
                itemProp="position"
              />
            </li>
            <li
              className="breadcrumbs__crumb"
              itemProp="itemListElement"
              itemScope={true}
              itemType="http://schema.org/ListItem"
              onClick={[Function]}
              role="presentation"
            >
              <a
                href="/guidance/indevelopment/gid-ng10186/consultation/"
                itemProp="item"
              >
                <span
                  itemProp="name"
                >
                  Consultation
                </span>
              </a>
              <meta
                content={1}
                itemProp="position"
              />
            </li>
          </ol>
        </nav>
        <main
          role="main"
        >
          Array [
            <h1
              className="page-header__heading mt--0"
              data-qa-sel="changeable-page-header"
            >
              Warren's Project
            </h1>,
            <div
              className="mb--d"
            >
              <p
                className="container container-full ml--0"
              >
                <span>
                  <span
                    className="tag tag--open"
                  >
                    Open for comments
                  </span>
                   Open until
                   
                  <time
                    dateTime={"2018-11-24T00:01:00.000Z"}
                  >
                    24 November 2018
                  </time>
                </span>
                  
                <a
                  href="/leadinformation"
                  onClick={[Function]}
                >
                  Request commenting lead permission
                </a>
              </p>
            </div>,
          ]
          <div
            className="grid"
          >
            <div
              data-g="12 md:3"
            >
              <nav
                aria-label="Chapters in this document"
                className="stacked-nav"
              >
                <h2
                  className="stacked-nav__root"
                >
                  Chapters in this document
                </h2>
                <ul
                  className="stacked-nav__list"
                >
                  <li
                    className="stacked-nav__list-item"
                    data-qa-sel="nav-list-item"
                  >
                    <a
                      aria-current="page"
                      href="/preview/GID-NG10186/consultation/113/document/1/chapter/draftappendicesa-tcleandocx"
                      onClick={[Function]}
                    >
                      <span
                        className="grid"
                      >
                        <span
                          className="pl--c"
                          data-g="10"
                        >
                          DRAFT_Appendices_A-T_CLEAN.docx
                        </span>
                      </span>
                    </a>
                  </li>
                  <li
                    className="stacked-nav__list-item"
                    data-qa-sel="nav-list-item"
                  >
                    <a
                      aria-current={null}
                      href="/preview/GID-NG10186/consultation/113/document/1/chapter/appendices"
                      onClick={[Function]}
                    >
                      <span
                        className="grid"
                      >
                        <span
                          className="pl--c"
                          data-g="10"
                        >
                          Appendices
                        </span>
                      </span>
                    </a>
                  </li>
                  <li
                    className="stacked-nav__list-item"
                    data-qa-sel="nav-list-item"
                  >
                    <a
                      aria-current={null}
                      href="/preview/GID-NG10186/consultation/113/document/1/chapter/scope"
                      onClick={[Function]}
                    >
                      <span
                        className="grid"
                      >
                        <span
                          className="pl--c"
                          data-g="10"
                        >
                          1 Scope
                        </span>
                      </span>
                    </a>
                  </li>
                </ul>
              </nav>
            </div>
            <div
              className="documentColumn"
              data-g="12 md:9"
            >
              <div
                className="document-comment-container "
              >
                <div
                  className="panel panel--inverse"
                >
                  <p>
                    <b>
                      There 
                      are 4 types of error
                       in this chapter.
                    </b>
                  </p>
                  <ul>
                    <li>
                      [E] - Error: This document contains neither a recognisable issue date nor a recognisable revision date, so appears to be completely undated
                    </li>
                    <li>
                      [E] - Error: No recognisable issue date found
                    </li>
                    <li>
                      [E] - Error: No main title found
                    </li>
                    <li>
                      [W] - Warning: Text found before first section heading, so software inserted section heading with title 'DRAFT_Appendices_A-T_CLEAN.docx'
                    </li>
                  </ul>
                </div>
                <div
                  className="chapter"
                  title="DRAFT_Appendices_A-T_CLEAN.docx"
                  xmlns="http://www.w3.org/1999/xhtml"
                >
                  <h2
                    className="title"
                  >
                    <a
                      className="annotator-chapter"
                      data-heading-type="chapter"
                      id="draftappendicesa-tcleandocx"
                      style={
                        Object {
                          "position": "relative",
                        }
                      }
                      title="Comment on chapter"
                      xmlns=""
                    >
                      DRAFT_Appendices_A-T_CLEAN.docx
                      <span
                        className="annotator-adder"
                      />
                    </a>
                    <div
                      className="ConversionError ConversionError--E"
                    >
                      <button
                        className="ConversionError__Button ConversionError__Button--E"
                        onClick={[Function]}
                      >
                        E
                      </button>
                      <ul
                        className="ConversionError__details ConversionError__details--E"
                      >
                        <li
                          className="ml--d"
                        >
                          This document contains neither a recognisable issue date nor a recognisable revision date, so appears to be completely undated
                        </li>
                      </ul>
                    </div>
                    <div
                      className="ConversionError ConversionError--E"
                    >
                      <button
                        className="ConversionError__Button ConversionError__Button--E"
                        onClick={[Function]}
                      >
                        E
                      </button>
                      <ul
                        className="ConversionError__details ConversionError__details--E"
                      >
                        <li
                          className="ml--d"
                        >
                          No recognisable issue date found
                        </li>
                      </ul>
                    </div>
                    <div
                      className="ConversionError ConversionError--E"
                    >
                      <button
                        className="ConversionError__Button ConversionError__Button--E"
                        onClick={[Function]}
                      >
                        E
                      </button>
                      <ul
                        className="ConversionError__details ConversionError__details--E"
                      >
                        <li
                          className="ml--d"
                        >
                          No main title found
                        </li>
                      </ul>
                    </div>
                    <div
                      className="ConversionError ConversionError--E"
                    >
                      <button
                        className="ConversionError__Button ConversionError__Button--E"
                        onClick={[Function]}
                      >
                        E
                      </button>
                      <ul
                        className="ConversionError__details ConversionError__details--E"
                      >
                        <li
                          className="ml--d"
                        >
                          This document contains neither a recognisable issue date nor a recognisable revision date, so appears to be completely undated
                        </li>
                      </ul>
                    </div>
                    <div
                      className="ConversionError ConversionError--E"
                    >
                      <button
                        className="ConversionError__Button ConversionError__Button--E"
                        onClick={[Function]}
                      >
                        E
                      </button>
                      <ul
                        className="ConversionError__details ConversionError__details--E"
                      >
                        <li
                          className="ml--d"
                        >
                          No recognisable issue date found
                        </li>
                      </ul>
                    </div>
                    <div
                      className="ConversionError ConversionError--E"
                    >
                      <button
                        className="ConversionError__Button ConversionError__Button--E"
                        onClick={[Function]}
                      >
                        E
                      </button>
                      <ul
                        className="ConversionError__details ConversionError__details--E"
                      >
                        <li
                          className="ml--d"
                        >
                          No main title found
                        </li>
                      </ul>
                    </div>
                    <div
                      className="ConversionError ConversionError--E"
                    >
                      <button
                        className="ConversionError__Button ConversionError__Button--E"
                        onClick={[Function]}
                      >
                        E
                      </button>
                      <ul
                        className="ConversionError__details ConversionError__details--E"
                      >
                        <li
                          className="ml--d"
                        >
                          This document contains neither a recognisable issue date nor a recognisable revision date, so appears to be completely undated
                        </li>
                      </ul>
                    </div>
                    <div
                      className="ConversionError ConversionError--E"
                    >
                      <button
                        className="ConversionError__Button ConversionError__Button--E"
                        onClick={[Function]}
                      >
                        E
                      </button>
                      <ul
                        className="ConversionError__details ConversionError__details--E"
                      >
                        <li
                          className="ml--d"
                        >
                          No recognisable issue date found
                        </li>
                      </ul>
                    </div>
                    <div
                      className="ConversionError ConversionError--E"
                    >
                      <button
                        className="ConversionError__Button ConversionError__Button--E"
                        onClick={[Function]}
                      >
                        E
                      </button>
                      <ul
                        className="ConversionError__details ConversionError__details--E"
                      >
                        <li
                          className="ml--d"
                        >
                          No main title found
                        </li>
                      </ul>
                    </div>
                    <div
                      className="ConversionError ConversionError--E"
                    >
                      <button
                        className="ConversionError__Button ConversionError__Button--E"
                        onClick={[Function]}
                      >
                        E
                      </button>
                      <ul
                        className="ConversionError__details ConversionError__details--E"
                      >
                        <li
                          className="ml--d"
                        >
                          This document contains neither a recognisable issue date nor a recognisable revision date, so appears to be completely undated
                        </li>
                      </ul>
                    </div>
                    <div
                      className="ConversionError ConversionError--E"
                    >
                      <button
                        className="ConversionError__Button ConversionError__Button--E"
                        onClick={[Function]}
                      >
                        E
                      </button>
                      <ul
                        className="ConversionError__details ConversionError__details--E"
                      >
                        <li
                          className="ml--d"
                        >
                          No recognisable issue date found
                        </li>
                      </ul>
                    </div>
                    <div
                      className="ConversionError ConversionError--E"
                    >
                      <button
                        className="ConversionError__Button ConversionError__Button--E"
                        onClick={[Function]}
                      >
                        E
                      </button>
                      <ul
                        className="ConversionError__details ConversionError__details--E"
                      >
                        <li
                          className="ml--d"
                        >
                          No main title found
                        </li>
                      </ul>
                    </div>
                    <div
                      className="ConversionError ConversionError--E"
                    >
                      <button
                        className="ConversionError__Button ConversionError__Button--E"
                        onClick={[Function]}
                      >
                        E
                      </button>
                      <ul
                        className="ConversionError__details ConversionError__details--E"
                      >
                        <li
                          className="ml--d"
                        >
                          This document contains neither a recognisable issue date nor a recognisable revision date, so appears to be completely undated
                        </li>
                      </ul>
                    </div>
                    <div
                      className="ConversionError ConversionError--E"
                    >
                      <button
                        className="ConversionError__Button ConversionError__Button--E"
                        onClick={[Function]}
                      >
                        E
                      </button>
                      <ul
                        className="ConversionError__details ConversionError__details--E"
                      >
                        <li
                          className="ml--d"
                        >
                          No recognisable issue date found
                        </li>
                      </ul>
                    </div>
                    <div
                      className="ConversionError ConversionError--E"
                    >
                      <button
                        className="ConversionError__Button ConversionError__Button--E"
                        onClick={[Function]}
                      >
                        E
                      </button>
                      <ul
                        className="ConversionError__details ConversionError__details--E"
                      >
                        <li
                          className="ml--d"
                        >
                          No main title found
                        </li>
                      </ul>
                    </div>
                    <div
                      className="ConversionError ConversionError--E"
                    >
                      <button
                        className="ConversionError__Button ConversionError__Button--E"
                        onClick={[Function]}
                      >
                        E
                      </button>
                      <ul
                        className="ConversionError__details ConversionError__details--E"
                      >
                        <li
                          className="ml--d"
                        >
                          This document contains neither a recognisable issue date nor a recognisable revision date, so appears to be completely undated
                        </li>
                      </ul>
                    </div>
                    <div
                      className="ConversionError ConversionError--E"
                    >
                      <button
                        className="ConversionError__Button ConversionError__Button--E"
                        onClick={[Function]}
                      >
                        E
                      </button>
                      <ul
                        className="ConversionError__details ConversionError__details--E"
                      >
                        <li
                          className="ml--d"
                        >
                          No recognisable issue date found
                        </li>
                      </ul>
                    </div>
                    <div
                      className="ConversionError ConversionError--E"
                    >
                      <button
                        className="ConversionError__Button ConversionError__Button--E"
                        onClick={[Function]}
                      >
                        E
                      </button>
                      <ul
                        className="ConversionError__details ConversionError__details--E"
                      >
                        <li
                          className="ml--d"
                        >
                          No main title found
                        </li>
                      </ul>
                    </div>
                    <div
                      className="ConversionError ConversionError--E"
                    >
                      <button
                        className="ConversionError__Button ConversionError__Button--E"
                        onClick={[Function]}
                      >
                        E
                      </button>
                      <ul
                        className="ConversionError__details ConversionError__details--E"
                      >
                        <li
                          className="ml--d"
                        >
                          This document contains neither a recognisable issue date nor a recognisable revision date, so appears to be completely undated
                        </li>
                      </ul>
                    </div>
                    <div
                      className="ConversionError ConversionError--E"
                    >
                      <button
                        className="ConversionError__Button ConversionError__Button--E"
                        onClick={[Function]}
                      >
                        E
                      </button>
                      <ul
                        className="ConversionError__details ConversionError__details--E"
                      >
                        <li
                          className="ml--d"
                        >
                          No recognisable issue date found
                        </li>
                      </ul>
                    </div>
                    <div
                      className="ConversionError ConversionError--E"
                    >
                      <button
                        className="ConversionError__Button ConversionError__Button--E"
                        onClick={[Function]}
                      >
                        E
                      </button>
                      <ul
                        className="ConversionError__details ConversionError__details--E"
                      >
                        <li
                          className="ml--d"
                        >
                          No main title found
                        </li>
                      </ul>
                    </div>
                    <div
                      className="ConversionError ConversionError--W"
                    >
                      <button
                        className="ConversionError__Button ConversionError__Button--W"
                        onClick={[Function]}
                      >
                        W
                      </button>
                      <ul
                        className="ConversionError__details ConversionError__details--W"
                      >
                        <li
                          className="ml--d"
                        >
                          Text found before first section heading, so software inserted section heading with title 'DRAFT_Appendices_A-T_CLEAN.docx'
                        </li>
                      </ul>
                    </div>
                  </h2>
                  <p />
                  <p />
                </div>
              </div>
            </div>
          </div>
        </main>
      </div>
    </div>
  </div>,
]
`;
